File: SharpMUSH.MarkupString/Markup/Markup.fs
namespace MarkupString

open System.Runtime.InteropServices
open System.Text.RegularExpressions
open ANSILibrary.ANSI
open System.Text.Json.Serialization

module MarkupImplementation =
  open ANSILibrary
  
  [<Struct>]
  type AnsiStructure =
    {
      Foreground: AnsiColor;
      Background: AnsiColor;
      LinkText: string option;
      LinkUrl: string option;
      Blink: bool;
      Bold: bool;
      Clear: bool;
      Faint: bool;
      Inverted: bool;
      Italic: bool;
      Overlined: bool;
      Underlined: bool;
      StrikeThrough: bool;
    }
    
  [<Struct>]
  type HtmlStructure =
    {
      TagName: string;
      Attributes: string option;
    }

  [<JsonDerivedType(typeof<NeutralMarkup>, "Neutral")>]
  [<JsonDerivedType(typeof<AnsiMarkup>, "Ansi")>]
  [<JsonDerivedType(typeof<HtmlMarkup>, "Html")>]
  type Markup =
      abstract member Wrap: string -> string
      abstract member WrapAndRestore: string * Markup -> string
      abstract member WrapAs: string * string -> string
      abstract member WrapAndRestoreAs: string * string * Markup -> string
      abstract member Prefix: string
      abstract member Postfix: string
      abstract member Optimize: string -> string
    
  and NeutralMarkup() =
    interface Markup with
      member this.Postfix: string = System.String.Empty
      member this.Prefix: string = System.String.Empty
      member this.Wrap(text: string): string = text
      member this.WrapAndRestore(text: string, _: Markup): string = text
      member this.WrapAs(_format: string, text: string): string = text
      member this.WrapAndRestoreAs(_format: string, text: string, _: Markup): string = text
      member this.Optimize(text: string): string = text

  and AnsiMarkup(details: AnsiStructure) =
    member val Details = details with get
    
    static member Create(
        [<Optional;DefaultParameterValue(null)>]?foreground, 
        [<Optional;DefaultParameterValue(null)>]?background, 
        [<Optional;DefaultParameterValue(null)>]?linkText, 
        [<Optional;DefaultParameterValue(null)>]?linkUrl, 
        [<Optional;DefaultParameterValue(false)>]?blink, 
        [<Optional;DefaultParameterValue(false)>]?bold, 
        [<Optional;DefaultParameterValue(false)>]?clear, 
        [<Optional;DefaultParameterValue(false)>]?faint, 
        [<Optional;DefaultParameterValue(false)>]?inverted, 
        [<Optional;DefaultParameterValue(false)>]?italic, 
        [<Optional;DefaultParameterValue(false)>]?overlined, 
        [<Optional;DefaultParameterValue(false)>]?underlined, 
        [<Optional;DefaultParameterValue(false)>]?strikeThrough) =
      {
        Foreground = defaultArg foreground AnsiColor.NoAnsi
        Background = defaultArg background AnsiColor.NoAnsi
        LinkText = defaultArg linkText (Some System.String.Empty)
        LinkUrl = defaultArg linkUrl (Some System.String.Empty)
        Blink = defaultArg blink false
        Bold = defaultArg bold false
        Clear = defaultArg clear false
        Faint = defaultArg faint false
        Inverted = defaultArg inverted false
        Italic = defaultArg italic false
        Overlined = defaultArg overlined false
        Underlined = defaultArg underlined false
        StrikeThrough = defaultArg strikeThrough false
      }
      |> AnsiMarkup

    /// Returns the CSS class names for the non-color formatting attributes.
    /// Only ms-* classes are returned; colors are rendered as inline style attributes.
    static member HtmlClassNames (details: AnsiStructure) : string list =
      [
        if details.Bold then yield "ms-bold"
        if details.Faint then yield "ms-faint"
        if details.Italic then yield "ms-italic"
        if details.Underlined then yield "ms-underline"
        if details.StrikeThrough then yield "ms-strike"
        if details.Overlined then yield "ms-overline"
        if details.Blink then yield "ms-blink"
      ]

    /// Converts an AnsiColor to a hex color string like "#ff0000".
    static member private colorToHex (color: AnsiColor) : string option =
      match color with
      | NoAnsi -> None
      | RGB c -> Some (sprintf "#%02x%02x%02x" c.R c.G c.B)
      | ANSI bytes ->
        let rgb = AnsiToRgb bytes
        Some (sprintf "#%02x%02x%02x" rgb.R rgb.G rgb.B)

    /// Renders an AnsiStructure as an HTML span.
    /// Colors are emitted as an inline style attribute; formatting flags are emitted as CSS classes.
    /// The <paramref name="text"/> parameter is expected to already be HTML-entity-encoded
    /// by the caller (getTextAs applies HtmlEncode to all Text leaf nodes before wrapping).
    static member private wrapAsHtmlClass (details: AnsiStructure) (text: string) : string =
      let fg, bg =
        if details.Inverted then details.Background, details.Foreground
        else details.Foreground, details.Background

      let colorStyle (property: string) (color: AnsiColor) =
        match AnsiMarkup.colorToHex color with
        | None -> None
        | Some hex -> Some (sprintf "%s: %s" property hex)

      let styles = [
        match colorStyle "color" fg with Some s -> yield s | None -> ()
        match colorStyle "background-color" bg with Some s -> yield s | None -> ()
      ]

      let classes = AnsiMarkup.HtmlClassNames details

      let inner =
        match details.LinkUrl with
        | Some url when url.Length > 0 ->
          sprintf "<a href=\"%s\">%s</a>" (System.Net.WebUtility.HtmlEncode url) text
        | _ -> text

      let styleAttr =
        if styles.IsEmpty then ""
        else sprintf " style=\"%s\"" (String.concat "; " styles)
      let classAttr =
        if classes.IsEmpty then ""
        else sprintf " class=\"%s\"" (String.concat " " classes)

      if styleAttr.Length = 0 && classAttr.Length = 0 then inner
      else sprintf "<span%s%s>%s</span>" styleAttr classAttr inner

    /// Renders an AnsiStructure as Pueblo-compatible HTML.
    /// Pueblo uses HTML 3.2-era tags: FONT COLOR for colors, B/I/U for formatting.
    /// The text parameter is expected to already be HTML-entity-encoded.
    static member wrapAsPueblo (details: AnsiStructure) (text: string) : string =
      let fg, bg =
        if details.Inverted then details.Background, details.Foreground
        else details.Foreground, details.Background

      text
      |> (fun t -> if details.StrikeThrough then sprintf "<S>%s</S>" t else t)
      |> (fun t -> if details.Overlined then sprintf "<SPAN STYLE=\"text-decoration: overline\">%s</SPAN>" t else t)
      |> (fun t -> if details.Underlined then sprintf "<U>%s</U>" t else t)
      |> (fun t -> if details.Italic then sprintf "<I>%s</I>" t else t)
      |> (fun t -> if details.Bold then sprintf "<B>%s</B>" t else t)
      |> (fun t -> match details.LinkUrl with
                   | Some url when url.Length > 0 -> sprintf "<A HREF=\"%s\">%s</A>" (System.Net.WebUtility.HtmlEncode url) t
                   | _ -> t)
      |> (fun t -> match AnsiMarkup.colorToHex bg with
                   | Some hex -> sprintf "<SPAN STYLE=\"background-color: %s\">%s</SPAN>" hex t
                   | None -> t)
      |> (fun t -> match AnsiMarkup.colorToHex fg with
                   | Some hex -> sprintf "<FONT COLOR=\"%s\">%s</FONT>" hex t
                   | None -> t)

    /// Renders an AnsiStructure as BBCode.
    /// Uses standard BBCode tags: [b], [i], [u], [s], [color=X].
    /// Text is passed through unchanged (BBCode does not use HTML encoding).
    static member wrapAsBBCode (details: AnsiStructure) (text: string) : string =
      let fg, _ =
        if details.Inverted then details.Background, details.Foreground
        else details.Foreground, details.Background

      text
      |> (fun t -> if details.StrikeThrough then sprintf "[s]%s[/s]" t else t)
      |> (fun t -> if details.Underlined then sprintf "[u]%s[/u]" t else t)
      |> (fun t -> if details.Italic then sprintf "[i]%s[/i]" t else t)
      |> (fun t -> if details.Bold then sprintf "[b]%s[/b]" t else t)
      |> (fun t -> match details.LinkUrl with
                   | Some url when url.Length > 0 -> sprintf "[url=%s]%s[/url]" url t
                   | _ -> t)
      |> (fun t -> match AnsiMarkup.colorToHex fg with
                   | Some hex -> sprintf "[color=%s]%s[/color]" hex t
                   | None -> t)

    /// Renders an AnsiStructure as MXP (MUD eXtension Protocol) tags.
    /// MXP uses HTML-like tags: <B>, <I>, <U>, <COLOR FORE=X BACK=Y>.
    /// The text parameter is expected to already be HTML-entity-encoded.
    static member wrapAsMxp (details: AnsiStructure) (text: string) : string =
      let fg, bg =
        if details.Inverted then details.Background, details.Foreground
        else details.Foreground, details.Background

      text
      |> (fun t -> if details.StrikeThrough then sprintf "<S>%s</S>" t else t)
      |> (fun t -> if details.Underlined then sprintf "<U>%s</U>" t else t)
      |> (fun t -> if details.Italic then sprintf "<I>%s</I>" t else t)
      |> (fun t -> if details.Bold then sprintf "<B>%s</B>" t else t)
      |> (fun t -> match details.LinkUrl with
                   | Some url when url.Length > 0 -> sprintf "<SEND HREF=\"%s\">%s</SEND>" (System.Net.WebUtility.HtmlEncode url) t
                   | _ -> t)
      |> (fun t -> match AnsiMarkup.colorToHex fg, AnsiMarkup.colorToHex bg with
                   | Some fh, Some bh -> sprintf "<COLOR FORE=\"%s\" BACK=\"%s\">%s</COLOR>" fh bh t
                   | Some fh, None -> sprintf "<COLOR FORE=\"%s\">%s</COLOR>" fh t
                   | None, Some bh -> sprintf "<COLOR BACK=\"%s\">%s</COLOR>" bh t
                   | None, None -> t)

    static member applyDetails (details: AnsiStructure) (text: string) =
        StringExtensions.toANSI text
        |> (fun t -> match details.LinkUrl with | None -> t | Some url -> if url.Length <> 0 then StringExtensions.linkANSI t url else t)
        |> (fun t -> match details.Foreground with | AnsiColor.NoAnsi -> t | fg -> StringExtensions.colorANSI t fg)
        |> (fun t -> match details.Background with | AnsiColor.NoAnsi -> t | bg -> StringExtensions.backgroundANSI t bg)
        |> (fun t -> if details.Blink then StringExtensions.blinkANSI(t) else t)
        |> (fun t -> if details.Bold then StringExtensions.boldANSI(t) else t)
        |> (fun t -> if details.Faint then StringExtensions.faintANSI(t) else t)
        |> (fun t -> if details.Italic then StringExtensions.italicANSI(t) else t)
        |> (fun t -> if details.Overlined then StringExtensions.overlinedANSI(t) else t)
        |> (fun t -> if details.Underlined then StringExtensions.underlinedANSI(t) else t)
        |> (fun t -> if details.StrikeThrough then StringExtensions.strikeThroughANSI(t) else t)
        |> (fun t -> if details.Inverted then StringExtensions.invertedANSI(t) else t)
        |> (fun t -> (if details.Clear then StringExtensions.clearANSI(t) else t))

    interface Markup with
      override this.Postfix: string = StringExtensions.endWithTrueClear(System.String.Empty).ToString()

      override this.Prefix: string = System.String.Empty

      // ANSI optimization moved to ANSILibrary.Optimization module
      [<TailCall>]
      override this.Optimize (text: string) : string =
        ANSILibrary.Optimization.optimize text

      override this.WrapAndRestore (text: string, outerDetails: Markup) : string =
        let restoreDetailsF (restoreDetails: Markup) =
          match restoreDetails with 
            | :? AnsiMarkup as markup -> 
              if details.Equals markup.Details then
                StringExtensions.toANSI System.String.Empty
              else
                StringExtensions.toANSI System.String.Empty
                |> (fun t -> match markup.Details.Foreground with | AnsiColor.NoAnsi -> t | fg -> StringExtensions.colorANSI t fg)
                |> (fun t -> match markup.Details.Background with | AnsiColor.NoAnsi -> t | bg -> StringExtensions.backgroundANSI t bg)
                |> (fun t -> if markup.Details.Blink then StringExtensions.blinkANSI(t) else t)
                |> (fun t -> if markup.Details.Bold then StringExtensions.boldANSI(t) else t)
                |> (fun t -> if markup.Details.Faint then StringExtensions.faintANSI(t) else t)
                |> (fun t -> if markup.Details.Italic then StringExtensions.italicANSI(t) else t)
                |> (fun t -> if markup.Details.Overlined then StringExtensions.overlinedANSI(t) else t)
                |> (fun t -> if markup.Details.Underlined then StringExtensions.underlinedANSI(t) else t)
                |> (fun t -> if markup.Details.StrikeThrough then StringExtensions.strikeThroughANSI(t) else t)
                |> (fun t -> if markup.Details.Inverted then StringExtensions.invertedANSI(t) else t)
            | :? HtmlMarkup -> StringExtensions.toANSI System.String.Empty // HTML tags don't need restoration codes
            | _ -> raise (System.Exception "Unknown markup type")
        (AnsiMarkup.applyDetails details text).ToString() + restoreDetailsF(outerDetails).ToString()

      override this.Wrap (text: string) : string =
        StringExtensions.endWithTrueClear((AnsiMarkup.applyDetails details text).ToString()).ToString()

      override this.WrapAs(format: string, text: string) : string =
        match format.ToLower() with
        | "html" -> AnsiMarkup.wrapAsHtmlClass details text
        | "pueblo" -> AnsiMarkup.wrapAsPueblo details text
        | "bbcode" -> AnsiMarkup.wrapAsBBCode details text
        | "mxp" -> AnsiMarkup.wrapAsMxp details text
        | _ -> (this :> Markup).Wrap(text)

      override this.WrapAndRestoreAs(format: string, text: string, outerDetails: Markup) : string =
        match format.ToLower() with
        | "html" -> AnsiMarkup.wrapAsHtmlClass details text
        | "pueblo" -> AnsiMarkup.wrapAsPueblo details text
        | "bbcode" -> AnsiMarkup.wrapAsBBCode details text
        | "mxp" -> AnsiMarkup.wrapAsMxp details text
        | _ -> (this :> Markup).WrapAndRestore(text, outerDetails)

  and HtmlMarkup(details: HtmlStructure) =
    member val Details = details with get
    
    static member Create(tagName: string, [<Optional;DefaultParameterValue(null)>]?attributes) =
      {
        TagName = tagName
        Attributes = defaultArg attributes None
      }
      |> HtmlMarkup

    interface Markup with
      // For HTML, we use empty prefix/postfix since Wrap handles everything
      override this.Postfix: string = System.String.Empty

      override this.Prefix: string = System.String.Empty

      override this.Wrap (text: string) : string =
        // Wrap includes the full HTML tags
        match details.Attributes with
        | None -> sprintf "<%s>%s</%s>" details.TagName text details.TagName
        | Some attrs -> sprintf "<%s %s>%s</%s>" details.TagName attrs text details.TagName

      override this.WrapAndRestore (text: string, outerDetails: Markup) : string =
        // For HTML, we just wrap without restoring outer markup since HTML tags are independent
        (this :> Markup).Wrap(text)

      override this.WrapAs(format: string, text: string) : string =
        match format.ToLower() with
        | "bbcode" -> text // HTML tags are not representable in BBCode; pass through text only
        | _ -> (this :> Markup).Wrap(text)

      override this.WrapAndRestoreAs(format: string, text: string, _: Markup) : string =
        match format.ToLower() with
        | "bbcode" -> text // HTML tags are not representable in BBCode; pass through text only
        | _ -> (this :> Markup).Wrap(text)

      override this.Optimize (text: string) : string = text

File: SharpMUSH.MarkupString/Markup/Markup.cs
// Converted from Markup.fs — namespace MarkupString.MarkupImplementation
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Serialization;
using ANSILibrary;

namespace MarkupString.MarkupImplementation;

// ── Struct records ─────────────────────────────────────────────────────────────

/// <summary>
/// Describes the ANSI/terminal formatting for an AnsiMarkup span.
/// </summary>
public readonly record struct AnsiStructure
{
    public AnsiColor Foreground    { get; init; }
    public AnsiColor Background    { get; init; }
    public string?  LinkText       { get; init; }
    public string?  LinkUrl        { get; init; }
    public bool     Blink          { get; init; }
    public bool     Bold           { get; init; }
    public bool     Clear          { get; init; }
    public bool     Faint          { get; init; }
    public bool     Inverted       { get; init; }
    public bool     Italic         { get; init; }
    public bool     Overlined      { get; init; }
    public bool     Underlined     { get; init; }
    public bool     StrikeThrough  { get; init; }
}

/// <summary>
/// Describes an HTML tag and optional attributes for an HtmlMarkup span.
/// </summary>
public readonly record struct HtmlStructure
{
    public string  TagName    { get; init; }
    public string? Attributes { get; init; }
}

// ── IMarkup interface ─────────────────────────────────────────────────────────

/// <summary>
/// Defines how to wrap a text segment in a given output format.
/// Implemented by NeutralMarkup, AnsiMarkup and HtmlMarkup.
/// </summary>
[JsonDerivedType(typeof(NeutralMarkup), "Neutral")]
[JsonDerivedType(typeof(AnsiMarkup),    "Ansi")]
[JsonDerivedType(typeof(HtmlMarkup),    "Html")]
public interface IMarkup
{
    string Wrap(string text);
    string WrapAndRestore(string text, IMarkup outerMarkup);
    string WrapAs(string format, string text);
    string WrapAndRestoreAs(string format, string text, IMarkup outerMarkup);
    string Prefix { get; }
    string Postfix { get; }
    string Optimize(string text);
}

// ── NeutralMarkup ─────────────────────────────────────────────────────────────

public sealed class NeutralMarkup : IMarkup
{
    public static readonly NeutralMarkup Instance = new();

    public string Prefix  => string.Empty;
    public string Postfix => string.Empty;
    public string Wrap(string text)                                     => text;
    public string WrapAndRestore(string text, IMarkup _)                => text;
    public string WrapAs(string _format, string text)                   => text;
    public string WrapAndRestoreAs(string _format, string text, IMarkup _) => text;
    public string Optimize(string text)                                 => text;
}

// ── AnsiMarkup ────────────────────────────────────────────────────────────────

public sealed class AnsiMarkup : IMarkup
{
    public AnsiStructure Details { get; }

    public AnsiMarkup(AnsiStructure details) => Details = details;

    /// <summary>Factory with all optional named parameters matching the F# Create signature.</summary>
    public static AnsiMarkup Create(
        AnsiColor? foreground  = null,
        AnsiColor? background  = null,
        string?    linkText    = null,
        string?    linkUrl     = null,
        bool       blink       = false,
        bool       bold        = false,
        bool       clear       = false,
        bool       faint       = false,
        bool       inverted    = false,
        bool       italic      = false,
        bool       overlined   = false,
        bool       underlined  = false,
        bool       strikeThrough = false)
    {
        return new AnsiMarkup(new AnsiStructure
        {
            Foreground    = foreground  ?? AnsiColor.NoAnsi.Instance,
            Background    = background  ?? AnsiColor.NoAnsi.Instance,
            LinkText      = linkText    ?? string.Empty,
            LinkUrl       = linkUrl     ?? string.Empty,
            Blink         = blink,
            Bold          = bold,
            Clear         = clear,
            Faint         = faint,
            Inverted      = inverted,
            Italic        = italic,
            Overlined     = overlined,
            Underlined    = underlined,
            StrikeThrough = strikeThrough,
        });
    }

    /// <summary>Returns CSS class names for non-color formatting attributes.</summary>
    public static IReadOnlyList<string> HtmlClassNames(AnsiStructure d)
    {
        var list = new List<string>();
        if (d.Bold)          list.Add("ms-bold");
        if (d.Faint)         list.Add("ms-faint");
        if (d.Italic)        list.Add("ms-italic");
        if (d.Underlined)    list.Add("ms-underline");
        if (d.StrikeThrough) list.Add("ms-strike");
        if (d.Overlined)     list.Add("ms-overline");
        if (d.Blink)         list.Add("ms-blink");
        return list;
    }

    private static string? ColorToHex(AnsiColor color) => color switch
    {
        AnsiColor.NoAnsi          => null,
        AnsiColor.RGB c           => $"#{c.Value.R:x2}{c.Value.G:x2}{c.Value.B:x2}",
        AnsiColor.ANSI bytes =>
            (ANSI.AnsiToRgb(bytes.Value)) is var rgb
                ? $"#{rgb.R:x2}{rgb.G:x2}{rgb.B:x2}"
                : null,
        _ => throw new NotSupportedException()
    };

    /// <summary>Renders AnsiStructure as an HTML span with inline style + CSS classes.</summary>
    public static string WrapAsHtmlClass(AnsiStructure d, string text)
    {
        var (fg, bg) = d.Inverted ? (d.Background, d.Foreground) : (d.Foreground, d.Background);

        var styles = new List<string>();
        if (ColorToHex(fg) is string fgHex) styles.Add($"color: {fgHex}");
        if (ColorToHex(bg) is string bgHex) styles.Add($"background-color: {bgHex}");

        var classes = HtmlClassNames(d);

        // Wrap hyperlink
        string inner = text;
        if (d.LinkUrl is { Length: > 0 } url)
            inner = $"<a href=\"{WebUtility.HtmlEncode(url)}\">{text}</a>";

        string styleAttr = styles.Count > 0 ? $" style=\"{string.Join("; ", styles)}\"" : "";
        string classAttr = classes.Count > 0 ? $" class=\"{string.Join(" ", classes)}\"" : "";

        return (styleAttr.Length == 0 && classAttr.Length == 0)
            ? inner
            : $"<span{styleAttr}{classAttr}>{inner}</span>";
    }

    public static string WrapAsPueblo(AnsiStructure d, string text)
    {
        var (fg, bg) = d.Inverted ? (d.Background, d.Foreground) : (d.Foreground, d.Background);
        string t = text;
        if (d.StrikeThrough) t = $"<S>{t}</S>";
        if (d.Overlined)     t = $"<SPAN STYLE=\"text-decoration: overline\">{t}</SPAN>";
        if (d.Underlined)    t = $"<U>{t}</U>";
        if (d.Italic)        t = $"<I>{t}</I>";
        if (d.Bold)          t = $"<B>{t}</B>";
        if (d.LinkUrl is { Length: > 0 } url)
            t = $"<A HREF=\"{WebUtility.HtmlEncode(url)}\">{t}</A>";
        if (ColorToHex(bg) is string bgHex) t = $"<SPAN STYLE=\"background-color: {bgHex}\">{t}</SPAN>";
        if (ColorToHex(fg) is string fgHex) t = $"<FONT COLOR=\"{fgHex}\">{t}</FONT>";
        return t;
    }

    public static string WrapAsBBCode(AnsiStructure d, string text)
    {
        var fg = d.Inverted ? d.Background : d.Foreground;
        string t = text;
        if (d.StrikeThrough) t = $"[s]{t}[/s]";
        if (d.Underlined)    t = $"[u]{t}[/u]";
        if (d.Italic)        t = $"[i]{t}[/i]";
        if (d.Bold)          t = $"[b]{t}[/b]";
        if (d.LinkUrl is { Length: > 0 } url)
            t = $"[url={url}]{t}[/url]";
        if (ColorToHex(fg) is string fgHex) t = $"[color={fgHex}]{t}[/color]";
        return t;
    }

    public static string WrapAsMxp(AnsiStructure d, string text)
    {
        var (fg, bg) = d.Inverted ? (d.Background, d.Foreground) : (d.Foreground, d.Background);
        string t = text;
        if (d.StrikeThrough) t = $"<S>{t}</S>";
        if (d.Underlined)    t = $"<U>{t}</U>";
        if (d.Italic)        t = $"<I>{t}</I>";
        if (d.Bold)          t = $"<B>{t}</B>";
        if (d.LinkUrl is { Length: > 0 } url)
            t = $"<SEND HREF=\"{WebUtility.HtmlEncode(url)}\">{t}</SEND>";
        string? fgHex = ColorToHex(fg);
        string? bgHex = ColorToHex(bg);
        t = (fgHex, bgHex) switch
        {
            ({ } f, { } b) => $"<COLOR FORE=\"{f}\" BACK=\"{b}\">{t}</COLOR>",
            ({ } f, null)  => $"<COLOR FORE=\"{f}\">{t}</COLOR>",
            (null, { } b)  => $"<COLOR BACK=\"{b}\">{t}</COLOR>",
            _              => t,
        };
        return t;
    }

    public static ANSIString ApplyDetails(AnsiStructure d, string text)
    {
        var s = text.ToANSI();
        if (d.LinkUrl is { Length: > 0 } url)  s = s.LinkANSI(url);
        if (d.Foreground is not AnsiColor.NoAnsi) s = s.ColorANSI(d.Foreground);
        if (d.Background is not AnsiColor.NoAnsi) s = s.BackgroundANSI(d.Background);
        if (d.Blink)          s = s.BlinkANSI();
        if (d.Bold)           s = s.BoldANSI();
        if (d.Faint)          s = s.FaintANSI();
        if (d.Italic)         s = s.ItalicANSI();
        if (d.Overlined)      s = s.OverlinedANSI();
        if (d.Underlined)     s = s.UnderlinedANSI();
        if (d.StrikeThrough)  s = s.StrikeThroughANSI();
        if (d.Inverted)       s = s.InvertedANSI();
        if (d.Clear)          s = s.ClearANSI();
        return s;
    }

    // ── IMarkup ────────────────────────────────────────────────────────────────

    public string Prefix  => string.Empty;
    public string Postfix => string.Empty.EndWithTrueClear().ToString();

    public string Optimize(string text) => Optimization.Optimize(text);

    public string Wrap(string text) =>
        ApplyDetails(Details, text).ToString().EndWithTrueClear().ToString();

    public string WrapAndRestore(string text, IMarkup outerMarkup)
    {
        var applied = ApplyDetails(Details, text).ToString();
        string restore = outerMarkup switch
        {
            AnsiMarkup am when am.Details.Equals(Details) =>
                string.Empty.ToANSI().ToString(),
            AnsiMarkup am =>
                BuildRestore(am.Details).ToString(),
            HtmlMarkup => string.Empty,
            _ => throw new Exception("Unknown markup type")
        };
        return applied + restore;
    }

    private static ANSIString BuildRestore(AnsiStructure d)
    {
        var s = string.Empty.ToANSI();
        if (d.Foreground is not AnsiColor.NoAnsi) s = s.ColorANSI(d.Foreground);
        if (d.Background is not AnsiColor.NoAnsi) s = s.BackgroundANSI(d.Background);
        if (d.Blink)          s = s.BlinkANSI();
        if (d.Bold)           s = s.BoldANSI();
        if (d.Faint)          s = s.FaintANSI();
        if (d.Italic)         s = s.ItalicANSI();
        if (d.Overlined)      s = s.OverlinedANSI();
        if (d.Underlined)     s = s.UnderlinedANSI();
        if (d.StrikeThrough)  s = s.StrikeThroughANSI();
        if (d.Inverted)       s = s.InvertedANSI();
        return s;
    }

    public string WrapAs(string format, string text) => format.ToLower() switch
    {
        "html"     => WrapAsHtmlClass(Details, text),
        "pueblo"   => WrapAsPueblo(Details, text),
        "bbcode"   => WrapAsBBCode(Details, text),
        "mxp"      => WrapAsMxp(Details, text),
        _          => Wrap(text),
    };

    public string WrapAndRestoreAs(string format, string text, IMarkup outerMarkup) =>
        format.ToLower() switch
        {
            "html"   => WrapAsHtmlClass(Details, text),
            "pueblo" => WrapAsPueblo(Details, text),
            "bbcode" => WrapAsBBCode(Details, text),
            "mxp"    => WrapAsMxp(Details, text),
            _        => WrapAndRestore(text, outerMarkup),
        };
}

// ── HtmlMarkup ────────────────────────────────────────────────────────────────

public sealed class HtmlMarkup : IMarkup
{
    public HtmlStructure Details { get; }

    public HtmlMarkup(HtmlStructure details) => Details = details;

    public static HtmlMarkup Create(string tagName, string? attributes = null) =>
        new(new HtmlStructure { TagName = tagName, Attributes = attributes });

    public static string WrapAsAnsi(HtmlStructure d, string text)
    {
        AnsiStructure? ansiDetails = d.TagName.ToLowerInvariant() switch
        {
            "b" or "strong"         => AnsiMarkup.Create(bold: true).Details,
            "i" or "em"             => AnsiMarkup.Create(italic: true).Details,
            "u"                     => AnsiMarkup.Create(underlined: true).Details,
            "s" or "strike" or "del"=> AnsiMarkup.Create(strikeThrough: true).Details,
            _                       => null,
        };
        return ansiDetails.HasValue
            ? AnsiMarkup.ApplyDetails(ansiDetails.Value, text).ToString().EndWithTrueClear().ToString()
            : text;
    }

    public string Prefix  => string.Empty;
    public string Postfix => string.Empty;

    public string Wrap(string text) => Details.Attributes is { } attrs
        ? $"<{Details.TagName} {attrs}>{text}</{Details.TagName}>"
        : $"<{Details.TagName}>{text}</{Details.TagName}>";

    public string WrapAndRestore(string text, IMarkup _) => Wrap(text);

    public string WrapAs(string format, string text) => format.ToLower() switch
    {
        "ansi"                          => WrapAsAnsi(Details, text),
        "bbcode" or "plaintext" or "plain" => text,
        _                               => Wrap(text),
    };

    public string WrapAndRestoreAs(string format, string text, IMarkup _) =>
        format.ToLower() switch
        {
            "ansi"                             => WrapAsAnsi(Details, text),
            "bbcode" or "plaintext" or "plain" => text,
            _                                  => Wrap(text),
        };

    public string Optimize(string text) => text;
}
